subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/ResourceSnapshotterCacheService.java

```diff
@@ -16,11 +16,19 @@
 
 package org.gradle.api.internal.changedetection.state;
 
+import org.gradle.internal.fingerprint.hashing.FileSystemLocationSnapshotHasher;
+import org.gradle.internal.fingerprint.hashing.RegularFileSnapshotContextHasher;
+import org.gradle.internal.fingerprint.hashing.RegularFileSnapshotContext;
 import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.snapshot.FileSystemLocationSnapshot;
 
 import javax.annotation.Nullable;
+import java.io.IOException;
 
 public interface ResourceSnapshotterCacheService {
     @Nullable
-    HashCode hashFile(RegularFileSnapshotContext fileSnapshotContext, RegularFileHasher hasher, HashCode configurationHash);
+    HashCode hashFile(FileSystemLocationSnapshot snapshot, FileSystemLocationSnapshotHasher hasher, HashCode configurationHash) throws IOException;
+
+    @Nullable
+    HashCode hashFile(RegularFileSnapshotContext fileSnapshotContext, RegularFileSnapshotContextHasher hasher, HashCode configurationHash) throws IOException;
```

Comment:
I think I now understand the difference between these two, and I think we could capture it in a more clear way.

IIUC, the difference is that the one taking `FileSystemLocationSnapshot` is expected to hash "root" input file entry as a whole, while the one taking a `RegularFileSnapshotContext` hashes some sort of "sub" entry, i.e. a file in a directory hierarchy, or a ZIP entry.

It would be good to find names for these concepts and then call these methods and the `RegularFileSnapshotContext` accordingly.

I think it might even make sense to remove `RegularFileSnapshotContext` and just pass the `relativePathSegmentSupplier` directly to the method here.

Reply:
Created https://github.com/gradle/gradle/issues/17653.